File: day_three/part_one.rb
class Pointer
  def initialize(start_x, start_y)
    @x = start_x
    @y = start_y
  end
  def up(amount=1)
    @y -= amount
  end
  def down(amount=1)
    @y += amount
  end
  def right(amount=1)
    @x += amount
  end
  def left(amount=1)
    @x -= amount
  end
  def get(line)
    if @x >= line.length() - 1
      @x = @x - (line.length() - 1)
    end
    return line[@x]
  end
end


class Frame
  def initialize()
    @lines = File.readlines("./input").each
  end
  def main()
    pointer = Pointer.new(0, 0)
    trees = 0
    @lines.each() do |line|
      item = pointer.get(line)
      if item == "#"
        trees += 1
      end
      pointer.down(1)
      pointer.right(3)
    end
    return trees
  end
end

frame = Frame.new()
trees = frame.main()
puts trees

File: day_three/part_one.py
class Pointer:
    def __init__(self, start_x, start_y):
        self.x = start_x
        self.y = start_y

    def up(self, amount: int = 1):
        self.y -= amount

    def down(self, amount: int = 1):
        self.y += amount

    def right(self, amount: int = 1):
        self.x += amount

    def left(self, amount: int = 1):
        self.x -= amount

    def get(self, lines):
        if self.x >= len(lines[self.y]) - 1:
            self.x = self.x - (len(lines[self.y]) - 1)
        return lines[self.y][self.x]


class Frame:
    def __init__(self):
        self.lines = self.get_lines()

    def get_lines(self):
        file = open("./input")
        return file.readlines()

    def main(self):
        pointer = Pointer(0, 0)
        self.trees = 0
        for lines in self.lines:
            item = pointer.get(self.lines)
            if item == "#":
                self.trees += 1
            pointer.down()
            pointer.right(3)

if __name__ == "__main__":
    frame = Frame()
    frame.main()
    print(frame.trees)
